### PR TITLE
Changed from internal allocation of buffers to external for better memory control

### PR DESCRIPTION
**Pull request by anedergaard**
- Pass buffers as spans to allow for external placement of buffers
- Allow difference read and write buffer sizes
- Added helpers to create a buffer of a type `HDLCPP` buffer type
   ```cpp
   template<size_t Capacity>
   using StaticBuffer = std::array<Container::value_type, Capacity>;
   ```
- Added helper to calculate overhead for HDLC
  ```cpp
  template<size_t Capacity>
    struct Calculate
    {
        // For details see: https://en.wikipedia.org/wiki/High-Level_Data_Link_Control#Structure
        static constexpr size_t Overhead = Capacity * 2 + 8;
    };
  ```
- Allows (again) dynamic allocation buffer size of hdlcpp object in python

### CODE REVIEW

File: include/Hdlcpp.hpp [hunk 1]
```diff
@@ -86,19 +89,30 @@ class Buffer {
...
+    using Container = std::span<uint8_t>;
 public:
+    template<size_t Capacity>
+    using StaticBuffer = std::array<Container::value_type, Capacity>;
+
+    template<size_t Capacity>
+    struct Calculate
+    {
+        // For details see: https://en.wikipedia.org/wiki/High-Level_Data_Link_Control#Structure
+        static constexpr size_t Overhead = Capacity * 2 + 8;
```
Comment (MIRBNO) [inline include/Hdlcpp.hpp:102]: IMO the names Capacity and Overhead are a bit misleading. Capacity is the original data size and Overhead is the encoded size (I would assume: Overhead = encoded size - original data size).

**Review by MIRBNO — APPROVED** [review]
Nice

File: include/Hdlcpp.hpp (hunk 1 above)
Comment (anedergaard) [reply include/Hdlcpp.hpp:102]: I agree that the names are not the best :disappointed: 
what about 
```cpp
static constexpr size_t WithOverhead = ...
```

Then creating a buffer would be like this
```cpp
static Hdlcpp::Hdlcpp::StaticBuffer<Hdlcpp::Hdlcpp::Calculate<HostProtocol::ReadBufferSize>::WithOverhead> readBuffer{};
```

Comment (anedergaard) [reply include/Hdlcpp.hpp:102]: These helpers should probably be moved out of the class so it is just scoped by the namespace
```cpp
static Hdlcpp::StaticBuffer<Hdlcpp::Calculate<HostProtocol::ReadBufferSize>::WithOverhead> readBuffer{};
```
Which makes it a bit more readable

Comment (MIRBNO) [reply include/Hdlcpp.hpp:102]: That works

File: README.md [hunk 2]
```diff
@@ -5,21 +5,21 @@ Hdlcpp is a header-only C++11 framing protocol optimized for embedded communicat
 
 ## Usage
 
-Hdlcpp requires that a transport read and write function is supplied as e.g. a [lambda expression](https://en.cppreference.com/w/cpp/language/lambda) for the actual data transport. Hereby Hdlcpp can easily be integrated e.g. with different UART implementations. It is also possible to increase the buffer size for encoding/decoding data, write timeout and number of write retries when constructing the instance.
+Hdlcpp requires that a transport read and write function is supplied as e.g. a [lambda expression](https://en.cppreference.com/w/cpp/language/lambda) for the actual data transport. Hereby Hdlcpp can easily be integrated e.g. with different UART implementations. It required a read and write buffer. The buffer type must to compatible with [std::span](https://en.cppreference.com/w/cpp/container/span), basically any contiguous sequence containers. The buffers can be size independent for encoding/decoding data, write timeout and number of write retries are also configurable when constructing the instance.
```
Comment (hansbinderup) [inline README.md:8]:
```suggestion
Hdlcpp requires that a transport read and write function is supplied as e.g. a [lambda expression](https://en.cppreference.com/w/cpp/language/lambda) for the actual data transport. Hereby Hdlcpp can easily be integrated e.g. with different UART implementations. It requires a read and write buffer. The buffer type must to compatible with [std::span](https://en.cppreference.com/w/cpp/container/span), basically any contiguous sequence containers. The buffers can be size independent for encoding/decoding data, write timeout and number of write retries are also configurable when constructing the instance.
```

**Review by hansbinderup — APPROVED** [review]
Lgtm :+1:

**Pull request merged**